### ISSUE TITLE
Possibility that smaller LM has better perplexity?

### ISSUE THREAD

**Issue by windweller** [kpu/kenlm#67]
Hi,

 I'm running KenLM on LM1B data (Language-Modeling 1 Billion), and for some weird reason, we see some perplexity goes down for extremely small model:

unigram tokens	unigram types	with OOV	exclude OOV
38023755 |	337972 |	156.5 |	148.27
3807417 |	107563 |	247.5 |	215.76
380918 |	32879 |	398.2 |	283.43
37438 |	8406 |	522.2 |	253.29
3728 |	1640 |	392.2 |	118.1

As you can notice that when unigram tokens drop low (smallest model), perplexity magically dropped to 392.2.

How does KenLM calculate perplexity including OOV and excluding OOV?

**Comment by kpu** [comment]
If you want to compare perplexities across different vocabularies, pick a big number (larger than the union of all vocabularies you are considering), and use 
`--vocab_pad BIG_NUMBER`
with `lmplz` when you train the model.  This will train a model which is normalized with respect to the vocabulary size you specify.  

The default normalizes with respect to whatever vocabulary exists, plus 1 for the unknown word.  

Remember to do the same with whatever you compare to: if you train a neural network on a 200,000-word vocabulary, then its vocabulary is normalized with respect to 200,001 words.  You should therefore take its unknown word probability times 1/(BIG_NUMBER-200,000) before computing perplexity.  

Remember, the model with no training data has OOV probability 1 and therefore perplexity 0.  Comparing perplexities with different vocabulary sizes is meaningless.  

You should really use including OOV, which is counts every word including unknown words.  Excluding OOV skips unknown words for purposes of computing probability and normalizing.  

Also, don't use the "1 billion word corpus".  They shuffled the sentences across training and test data.  So if there's a paragraph that talks about FooBarQuuxes, there's a much higher chance than normal that some will be in the training data for each test sentence.  Real evaluation sets shuffle paragraphs or documents at a time for this reason.

**Comment by windweller** [comment]
Thank you for responding. That spawns into two smaller questions:

1. Which database would you recommend that is large enough but we can also learn a meaningful language model?
2. When using `kenlm/bin/query` to compute perplexity on a validation set, does the program use line breaks to compute perplexity? (Perplexity per line then aggregate and average them?)
(The reason I asked is because when running it on a file with no line breaks, perplexity gets ridiculously low - like 0.7 or something...)

**Comment by kpu** [comment]
1. Pick some training corpora from http://statmt.org/wmt16/translation-task.html then use a test set for perplexity http://data.statmt.org/wmt16/translation-task/test.tgz .  Note that the "billion word corpus" is actually a subset of this data; you're just getting it before their weird sampling procedure.  Concatenating all the newscrawls then testing on WMT16's test set would be reasonable.  

2. I hadn't really designed the query program for a line without a proper line-break at the very end: 
https://github.com/kpu/kenlm/blob/master/lm/ngram_query.hh#L82
I'll need to fix the accounting in that case.

Line breaks implicitly append the end of sentence marker (as a word) then start conditioning on the beginning of sentence.  This can be disabled with `-n`, in which case it will start with empty context (which is not appropriate for Kneser-Ney smoothing).  The query program also uses line breaks for numerical precision purposes, accumulating

**Comment by windweller** [comment]
This has been so helpful! Thank you!!

**Comment by muhammadhermawan** [comment]
@kpu mr heafield can you help me for showing the command line to compute perplexity in kenlm ?
